client/state/data-layer/wpcom/sites/plugins/new/index.js

```diff
@@ -0,0 +1,51 @@
+/**
+ * Internal dependencies
+ */
+import { PLUGIN_UPLOAD, PLUGIN_INSTALL_REQUEST_SUCCESS } from 'state/action-types';
+import {
+	completePluginUpload,
+	pluginUploadError,
+	updatePluginUploadProgress,
+} from 'state/plugins/upload/actions';
+import { dispatchRequest } from 'state/data-layer/wpcom-http/utils';
+import { http } from 'state/data-layer/wpcom-http/actions';
+
+const uploadPlugin = ( { dispatch }, action ) => {
+	const { siteId, file } = action;
+
+	dispatch( http( {
+		method: 'POST',
+		path: `/sites/${ siteId }/plugins/new`,
+		apiVersion: '1',
+		formData: [ [ 'zip[]', file ] ],
+	}, action ) );
+};
+
+export const uploadComplete = ( { dispatch }, { siteId }, next, data ) => {
+	const { id: pluginId } = data;
+	dispatch( completePluginUpload( siteId, pluginId ) );
+	dispatch( {
+		type: PLUGIN_INSTALL_REQUEST_SUCCESS,
+		siteId,
+		pluginId,
+		data
+	} );
+};
+
+export const receiveError = ( { dispatch }, { siteId }, next, error ) => {
+	dispatch( pluginUploadError( siteId, error ) );
```

Comment:
The reason for dispatching `PLUGIN_INSTALL_REQUEST_SUCCESS` is to attempt to take advantage of the existing logic for adding the details of a newly installed plugin to local state. I'm not 100% sure this will actually work, it may need some tweaking.

Either way, I don't thing we want to dispatch any other related actions because we don't want to install a plugin in the usual way (download from .org).